src/dmd/dmangle.d

```diff
@@ -543,9 +543,26 @@ public:
         }
     }
 
+    /************************************************************
+     * Try to obtain an externally mangled identifier from a declaration.
+     * If the declaration is at global scope or mixed in at global scope,
+     * the user might want to call it externally, so an externally mangled
+     * name is returned. Member functions or nested functions can't be called
+     * externally in C, so in that case null is returned. C++ does support
+     * namespaces, so extern(C++) always gives a C++ mangled name.
+     *
+     * See also: https://issues.dlang.org/show_bug.cgi?id=20012
+     *
+     * Params:
+     *     d = declaration to mangle
+     *
+     * Returns:
+     *     an externally mangled name or null if the declaration cannot be called externally
+     */
     extern (D) static const(char)[] externallyMangledIdentifier(Declaration d)
     {
-        if (!d.parent || d.parent.isModule() || d.linkage == LINK.cpp) // if at global scope
+        const par = d.toParent(); //toParent() skips over mixin templates
+        if (!par || par.isModule() || d.linkage == LINK.cpp)
```

Comment:
Is this actually a problem of mixins? Why would extern(C) still behave differently than extern(C++) everywhere else? I remember wondering why I cannot add an extern(C) function declaration (without definition) in a local scope just to call it (that also doesn't work for C++, though, as the function name is added as a namespace).

Reply:
> Is this actually a problem of mixins?

It has to do with whether the declaration is accessible from global scope. With D's current feature set, that means either being directly there or mixed in at global scope.

> Why would extern(C) still behave differently than extern(C++) everywhere else?

Because C++ has namespaces and C does not.

> I remember wondering why I cannot add an extern(C) function declaration (without definition) in a local scope just to call it

Because C has no nested functions / static functions in function scope.
There was a proposal to completely invalidate extern(C) in nested scopes, but it was rejected:
https://github.com/dlang/dmd/pull/7229
This PR aims to solve a specific use case (mixin templates for C boilerplate) and leave the rest alone. It's possible to follow this up with a change in the behavior of extern(C) in nested scopes or extern(C++), but for now this should be an improvement on its own.

Reply:
C++ doesn't have local functions >_<

Reply:
> There was a proposal to completely invalidate extern(C) in nested scopes, but it was rejected: #7229

Ah, ok. Thanks for the link. I was also wondering whether widening the scope of the PR would break too much code.

> C++ doesn't have local functions >_<

This compiles on godbolt;
```
auto foo(int)
{
    struct S
    {
        static void foo() {}
    };
    return &S::foo;
}
```

Reply:
That's not quite a local function, that's a method of a local class.
How does that mangle?

Reply:
> How does that mangle?

_ZZ3fooiEN1S3fooEv demangling to foo(int)::S::foo()

The corresponding D code doesn't add the function arguments to the extern(C++) mangling leading to issues as described in the third item above.

Reply:
Okay, interesting. We should get some unit tests to prove that works. I never thought to test it when I was tinkering with that stuff.